Sources/NIO/ConvenienceOptionSupport.swift

```diff
@@ -0,0 +1,440 @@
+        self.data = data
+    }
+    
+    /// Apply the contained option to the supplied object (almost certainly bootstrap) using the default mapping.
+    /// - Parameter with: object to use to apply the option.
+    /// - Returns: the modified object
+    public func applyOptionDefaultMapping<OptionApplier : NIOChannelOptionAppliable>(with optionApplier: OptionApplier) -> OptionApplier {
+        return data.applyOptionDefaultMapping(with: optionApplier)
+    }
+    
+    fileprivate enum ShorthandOption {
+        case reuseAddr
+        case disableAutoRead
+        case allowRemoteHalfClosure(Bool)
+        
+        func applyOptionDefaultMapping<OptionApplier : NIOChannelOptionAppliable>(with optionApplier: OptionApplier) -> OptionApplier {
+            switch self {
+            case .reuseAddr:
+                return optionApplier.applyOption(ChannelOptions.socketOption(.reuseaddr), value: 1)
+            case .allowRemoteHalfClosure(let value):
+                return optionApplier.applyOption(ChannelOptions.allowRemoteHalfClosure, value: value)
+            case .disableAutoRead:
+                return optionApplier.applyOption(ChannelOptions.autoRead, value: false)
+            }
+        }
+    }
+}
+
+// Hashable for the convenience of users.
+extension NIOTCPShorthandOption: Hashable {}
+extension NIOTCPShorthandOption.ShorthandOption: Hashable {}
+
+/// Approved shorthand options.
+extension NIOTCPShorthandOption {
+    /// Allow immediately reusing a local address.
+    public static let allowImmediateEndpointAddressReuse = NIOTCPShorthandOption(.reuseAddr)
+    
+    /// The user will manually call `Channel.read` once all the data is read from the transport.
+    public static let disableAutoRead = NIOTCPShorthandOption(.disableAutoRead)
+    
+    /// Allows users to configure whether the `Channel` will close itself when its remote
+    /// peer shuts down its send stream, or whether it will remain open. If set to `false` (the default), the `Channel`
+    /// will be closed automatically if the remote peer shuts down its send stream. If set to true, the `Channel` will
+    /// not be closed: instead, a `ChannelEvent.inboundClosed` user event will be sent on the `ChannelPipeline`,
+    /// and no more data will be received.
+    public static let allowRemoteHalfClosure =
+        NIOTCPShorthandOption(.allowRemoteHalfClosure(true))
+    
+    /// Allows users to configure whether the `Channel` will close itself when its remote
+    /// peer shuts down its send stream, or whether it will remain open. If set to `false` (the default), the `Channel`
+    /// will be closed automatically if the remote peer shuts down its send stream. If set to true, the `Channel` will
+    /// not be closed: instead, a `ChannelEvent.inboundClosed` user event will be sent on the `ChannelPipeline`,
+    /// and no more data will be received.
+    public static func allowRemoteHalfClosure(_ value: Bool) -> NIOTCPShorthandOption {
```

Comment:
why do we have the no argument as well as the `value: Bool` option here but for the reuse address case we don't?

Reply:
Because it's used like this in various examples.  Happy to make it consistent in whichever way popular opinion likes.

Reply:
I'd like consistency but currently not sure which one I want :P. CC @Lukasa / @glbrntt

Reply:
I'm disinclined to have the with-bool versions. Just the plain names.

Reply:
+1 on just plain names

Reply:
Killed the bool versions.